CMakeLists.txt

```diff
@@ -0,0 +1,197 @@
+cmake_minimum_required (VERSION 3.0)
+project (s2n)
+##header files
+file(GLOB API_HEADERS
+     "api/*.h"
+)
+
+file(GLOB CRYPTO_HEADERS
+    "crypto/*.h"
+)
+
+file(GLOB ERROR_HEADERS
+    "error/*.h"
+)
+
+file(GLOB STUFFER_HEADERS
+    "stuffer/*.h"
+)
+
+file(GLOB TLS_HEADERS
+    "tls/*.h"
+)
+
+file(GLOB UTILS_HEADERS
+    "utils/*.h"
+)
+ 
+ ##source files
+ file(GLOB CRYPTO_SRC
+     "crypto/*.c"
+ )
+
+file(GLOB ERROR_SRC
+    "error/*.c"
+)
+
+file(GLOB STUFFER_SRC
+    "stuffer/*.c"
+)
+
+file(GLOB TLS_SRC
+    "tls/*.c"
+)
+
+file(GLOB UTILS_SRC
+    "utils/*.c"
+)
+
+##be nice to visual studio users
+if(MSVC)
+    source_group("Header Files\\s2n\\api" FILES ${API_HEADERS})
+    source_group("Header Files\\s2n\\crypto" FILES ${CRYPTO_HEADERS})
+    source_group("Header Files\\s2n\\error" FILES ${ERROR_HEADERS})
+    source_group("Header Files\\s2n\\stuffer" FILES ${STUFFER_HEADERS})
+    source_group("Header Files\\s2n\\tls" FILES ${TLS_HEADERS})
+    source_group("Header Files\\s2n\\utils" FILES ${UTILS_HEADERS})
+
+    source_group("Source Files\\crypto" FILES ${CRYPTO_SRC})
+    source_group("Source Files\\error" FILES ${ERROR_SRC})
+    source_group("Source Files\\stuffer" FILES ${STUFFER_SRC})
+    source_group("Source Files\\tls" FILES ${TLS_SRC})
+    source_group("Source Files\\utils" FILES ${UTILS_SRC})
+endif()
+
+if(APPLE)
+    set(OS_LIBS c pthread)        
+elseif(CMAKE_SYSTEM_NAME STREQUAL "FreeBSD")
+    set(OS_LIBS thr)
+elseif(CMAKE_SYSTEM_NAME STREQUAL "NetBSD")
+    set(OS_LIBS pthread)
+else()
+    set(OS_LIBS pthread dl rt)
+endif()
+ 
+file(GLOB S2N_HEADERS
+    ${API_HEADERS}
+    ${CRYPTO_HEADERS}
+    ${ERROR_HEADERS}
+    ${STUFFER_HEADERS}
+    ${TLS_HEADERS}
+    ${UTILS_HEADERS}
+ )
+ 
+ file(GLOB S2N_SRC
+    ${CRYPTO_SRC}
+    ${ERROR_SRC}
+    ${STUFFER_SRC}
+    ${TLS_SRC}
+    ${UTILS_SRC}
+ )
+ 
+if(BUILD_SHARED_LIBS)
+    add_library(s2nShared SHARED ${S2N_HEADERS} ${S2N_SRC})
+    set_target_properties(s2nShared PROPERTIES OUTPUT_NAME s2n)
+    set_target_properties(s2nShared PROPERTIES LINKER_LANGUAGE C) 
+endif()
+
+add_library(s2nStatic STATIC ${S2N_HEADERS} ${S2N_SRC})
+set_target_properties(s2nStatic PROPERTIES OUTPUT_NAME s2n)
+set_target_properties(s2nStatic PROPERTIES LINKER_LANGUAGE C)
```

Comment:
@JonathanHenson if the only difference between the static and the dynamic builds is the name of the output library, you can make the name of the library a variable and eliminate lots of the duplication in the file.

Reply:
To be a good CMake citizen we should respect the command line arguments passed by the user.
In other words, **always** building the static version is a surprising behavior. The value of ``BUILD_SHARED_LIBS`` should be what controls the type of library output.